crates/rolldown/src/stages/link_stage/tree_shaking.rs

```diff
@@ -23,16 +23,23 @@ fn include_module(ctx: &mut Context, module: &NormalModule) {
   }
   ctx.is_module_included_vec[module.id] = true;
 
-  if !matches!(module.side_effects, DeterminedSideEffects::NoTreeshake)
-    && (ctx.tree_shaking || module.id == ctx.runtime_id)
-  {
+  if module.id == ctx.runtime_id {
+    // runtime module has no side effects and it's statements should be included
+    // by other modules's references.
+    return;
+  }
+
+  if ctx.tree_shaking {
+    let forced_no_treeshake = matches!(module.side_effects, DeterminedSideEffects::NoTreeshake);
     module.stmt_infos.iter_enumerated().for_each(|(stmt_info_id, stmt_info)| {
-      if stmt_info.side_effect {
+      if stmt_info.side_effect && !forced_no_treeshake {
```

Comment:
Any good of this? you need to calculate every time, we only calculate once previous.

Reply:
I want make it more clear to see the different between enabling treeshaking or not. But yes bad for performance. Let's do a better way.